datasette/cli.py

```diff
@@ -52,22 +52,34 @@ def publish(publisher, files, name, metadata, extra_options, force, **extra_meta
 
     Example usage: datasette publish now my-database.db
     """
-    if not shutil.which('now'):
-        click.secho(
-            ' The publish command requires "now" to be installed and configured ',
-            bg='red',
-            fg='white',
-            bold=True,
-            err=True,
-        )
-        click.echo('Follow the instructions at https://zeit.co/now#whats-now', err=True)
-        sys.exit(1)
-
-    with temporary_docker_directory(files, name, metadata, extra_options, extra_metadata):
-        if force:
-            call(['now', '--force'])
-        else:
-            call('now')
+    if publisher == 'now':
+        if not shutil.which('now'):
+            click.secho(
+                ' The publish command requires "now" to be installed and configured ',
+                bg='red',
+                fg='white',
+                bold=True,
+                err=True,
+            )
+            click.echo('Follow the instructions at https://zeit.co/now#whats-now', err=True)
+            sys.exit(1)
+
+        with temporary_docker_directory(files, name, metadata, extra_options, extra_metadata):
+            if force:
+                call(['now', '--force'])
+            else:
+                call('now')
+
+    elif publisher == 'heroku':
+        # FIXME: need to verify we have heroku, heroku-builds, and are logged in (ugh)
+        with temporary_heroku_directory(files, name, metadata, extra_options, extra_metadata):
+            # build(files) doesn't work, dunno why
+            app = Datasette(files)
+            open("inspect-data.json", 'w').write(json.dumps(app.inspect(), indent=2))
+
+            create_output = check_output(['heroku', 'apps:create', '--json'])
+            app_name = json.loads(create_output)["name"]
+            call(["heroku", "builds:create", "-a", app_name])
```

Comment:
Yeah - do we want to force git to be installed, though?

Reply:
Huh good question. Datasette doesn't work on Windows (a Sanic restriction) and git ships with OS X these days so I don't see it as a particularly nasty limitation. We can show an error if someone tries to run this and they don't have git.

Reply:
So, having messed with this a bit, there might be a good reason to stick to the builds API: doing the git approach is super-messy when it comes to updating the app. Either you have to `git push --force` each time, which eliminates history in the remote, or you have to get super-clever about cloning a remote repo first if the app's already been pushed before in the past.

So I see 3 options:

1. Use the git approach anyway, and just deal with complexity. Upside: no additional plugins or API calls. Downside: messy code, lots of conditionals, potentially slow cloning existing apps.

2. Use [heroku-builds](https://github.com/heroku/heroku-builds), and just prompt for (or auto-install?) the plugin. Upside: simple code. Downside: plugin, potentially brittle there.

3. Call the build API directly (using requests or whatever). Upside: no other binaries needed (would even work w/o the Heroku CLI being installed if we wanted to jump that hoop). Downside: lots more code in datasette to deal with that complexity.

Personally, I'm leaning towards option 2: keep using the builds plugin, and just offer to install it for folks who don't have it. 

@simonw how'd you like to proceed?

Reply:
I'm happy with option 2 then - heroku plug-ins are pretty painless.

I've not had to think about updates for the Zeit Now and docker package options because they are both meant to be immutable containers that you replace entirely every time you ship code.

I have been thinking about adding an option that lets you "persist" the temporary directory created by "datasette now publish" in case you want to manually customize it or keep it around - something like this:

    datasette publish now mydatabase.db --persist-directory=~/my-project